### PR TITLE
Objects and Object Constructors: Update lesson on Prototype and Prototypal Inheritance

### PR DESCRIPTION
**Pull request by anupamvamsi**
## Because
The primary aim of this PR was to update the lesson's Recommended Method for Prototypal Inheritance content from using `Object.create()` to the `Object.setPrototypeOf()` method.

However I realized that the lesson also could better explain the `prototype` function property vs. the `__proto__` / `[[Prototype]]` property and prototypal inheritance through TOP's curriculum directly. Personally to me as a learner, the lesson was confusing and I had to refer to quite a few resources to understand it.

It is a complete update to #24560 and its corresponding PRs.

## This PR
1. Explains the concept of the `prototype` and Prototypal Inheritance.
2. Updates the recommended way of performing prototypal inheritance to using `Object.setPrototypeOf()` rather than using `Object.create()`.

## Issue
Closes #24560, #24563, #24782.

## Additional Information
* This PR has been made as per suggestions on the conversation on #24560.
* Suggestions-Bugs Forum discussion on [Discord](https://discord.com/channels/505093832157691914/1010249197544538112)
* Discussion on mdn/content: https://github.com/mdn/content/pull/18880
* PR closing the above discussion on mdn/content: https://github.com/mdn/content/pull/20693

## Pull Request Requirements
-   [x] I have thoroughly read and understand [The Odin Project Contributing Guide](https://github.com/TheOdinProject/theodinproject/blob/main/CONTRIBUTING.md)
-   [x] The title of this PR follows the `location of change: brief description of change` format, e.g. `Intro to HTML and CSS lesson: Fix link text`
-   [x] The `Because` section summarizes the reason for this PR
-   [x] The `This PR` section has a bullet point list describing the changes in this PR
-   [x] If this PR addresses an open issue, it is linked in the `Issue` section
-   [x] If any lesson files are included in this PR, they have been previewed with the [Markdown preview tool](https://www.theodinproject.com/lessons/preview) to ensure it is formatted correctly
-   [x] If any lesson files are included in this PR, they follow the [Layout Style Guide](https://github.com/TheOdinProject/curriculum/blob/main/LAYOUT_STYLE_GUIDE.md)

### CODE REVIEW

**Comment by anupamvamsi** [comment]
Hi, @wise-king-sullyman! I've made a PR for the complete set of changes as you suggested. Please take a look, thanks! 

I'll be closing the other PRs I've created before.

File: javascript/organizing_your_javascript_code/objects_and_object_constructors.md [hunk 1]
```diff
@@ -149,94 +149,204 @@ console.log(theHobbit.info());
...
+
+Essentially, this is how JavaScript makes use of `prototype` and `__proto__` - by having the `__proto__` values of objects point to `prototype`s and inheriting from those prototypes, and thus forming a chain. This kind of inheritance using prototypes is hence named as Prototypal inheritance. JavaScript figures out which properties exist (or do not exist) on the object and starts traversing the chain to find the property or function, like so:
+
+1. Is the `.valueOf` function part of the `player1` object? No, it is not. (Remember, only the `name`, `marker` and `sayName` properties are part of the `Player` objects.)
+
+2. Is the function part of the `player1`'s prototype (the `player1.__proto__` value, i.e., `Player.prototype`)? No, only the `.sayHello` function is a part of it.
+
+3. Well, then, is it part of `Player.prototype.__proto__` (=== `Object.prototype`)? Yes, `.valueOf` is defined on `Object.prototype`!
```
Comment (fortypercenttitanium) [inline javascript/organizing_your_javascript_code/objects_and_object_constructors.md:253]: Can we use this aside to also cover how `hasOwnProperty` works as well? It's in a very similar vein and is a method that is likely to be used or encountered.

File: javascript/organizing_your_javascript_code/objects_and_object_constructors.md [hunk 2]
```diff
@@ -149,94 +149,204 @@ console.log(theHobbit.info());
...
+Note:
+
+1. Every `prototype` object inherits from `Object.prototype` by default.
+
+2. An object's `.__proto__` property can only point to _one_ `prototype` object at a time. 
 
 #### Recommended Method for Prototypal Inheritance
 
-So far you have seen several ways of making an object inherit the prototype from another object. At this point in history, the recommended way of setting the prototype of an object is `Object.create` ([here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/create) is the documentation for that method). `Object.create` very simply returns a new object with the specified prototype and any additional properties you want to add. For our purposes, you use it like so:
+Now, how do you utilise Prototypal Inheritance? What do you need to do to use it? Just as we use `Object.getPrototypeOf()` to 'get' or view the `prototype` of an object, we can use `Object.setPrototypeOf()` to 'set' or mutate it. Let's see how it works by adding a `Person` Object Constructor to the `Player` example, and making `Player` inherit from `Person`!
```
Comment (fortypercenttitanium) [inline javascript/organizing_your_javascript_code/objects_and_object_constructors.md:265]:
```suggestion
Now, how do you utilize Prototypal Inheritance? What do you need to do to use it? Just as we use `Object.getPrototypeOf()` to 'get' or view the `prototype` of an object, we can use `Object.setPrototypeOf()` to 'set' or mutate it. Let's see how it works by adding a `Person` Object Constructor to the `Player` example, and making `Player` inherit from `Person`!
```

File: javascript/organizing_your_javascript_code/objects_and_object_constructors.md [hunk 3]
```diff
@@ -149,94 +149,204 @@ console.log(theHobbit.info());
...
+   
+   2b. The _reference_ to this value of `Player.prototype` is stored in every `Player` object's `.__proto__` property, every time a `Player` object is created. 
+   
+   2c. Hence, you get a `true` value returned when you compute `player1.__proto__ === Player.prototype`.
+
+3. **...that the original object _inherits_ from, and has access to all of its prototype's methods and properties**: 
+
+   3a. As said in the earlier point, every `Player` object's `.__proto__` property refers to `Player.prototype`: `player1.__proto__ === player2.__proto__` (returns `true`).
+   
+   3b. So, any properties or methods defined on `Player.prototype` will be available to the created `Player` objects!
```
Comment (fortypercenttitanium) [inline javascript/organizing_your_javascript_code/objects_and_object_constructors.md:197]: Please refer to the style guide on how we number our markdown [ordered lists (lazy numbering) ](https://github.com/TheOdinProject/curriculum/blob/main/LAYOUT_STYLE_GUIDE.md#lazy-numbering)

File: javascript/organizing_your_javascript_code/objects_and_object_constructors.md [hunk 4]
```diff
@@ -149,94 +149,204 @@ console.log(theHobbit.info());
...
+
+~~~javascript
+Object.getPrototypeOf(player1) === Player.prototype // returns true
+Object.getPrototypeOf(player2) === Player.prototype // returns true
 ~~~
 
-If you're using constructors to make your objects it is best to define functions on the `prototype` of that object. Doing so means that a single instance of each function will be shared between all of the Student objects. If we declare the function directly in the constructor, like we did when they were first introduced, that function would be duplicated every time a new Student is created. In this example, that wouldn't really matter much, but in a project that is creating thousands of objects, it really can make a difference.
+In some places you might also come across `[[Prototype]]`, which is just another way of talking about the `.__proto__` property of an object, like `player1.[[Prototype]]`.
+
+This explanation about the `prototype` might have been a lot, so remember to take a breather before moving on!
```
Comment (fortypercenttitanium) [inline javascript/organizing_your_javascript_code/objects_and_object_constructors.md:223]: Use this section to make a note about `__proto__` and `[[Prototype]]`, mention how you might see them in legacy code, etc... See the bigger comment on the review to understand why we prefer not teaching to use `__proto__` in lieu of `Object.getPrototypeOf()`

**Review by fortypercenttitanium — CHANGES_REQUESTED** [review]
Thank you for your patience on this one. We had a lot of internal discussion about this and have decided that this material is a good addition to the curriculum, however, we'd like to graciously ask for a few modifications:

- Prefer using `Object.getPrototypeOf` in lieu of __proto__ as the primary method for getting an object's prototype. Since [__proto__ is deprecated](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/proto), we don't want to be using it as a tool for learning modern Javascript.
- A couple of style guide conformity changes notated on the appropriate lines
- Please use the aside section on current lines 212-223 to make mention of the deprecated methods of viewing prototype info.
- Please make mention of the `Object.hasOwnProperty` method and how it relates to the prototype chain details you are explaining at that point.

**Comment by anupamvamsi** [comment]
Hi @fortypercenttitanium, thanks for your inputs! I understand that this is a topic of debate and it required some time from your side to be spent on it to decide if it is good for the lesson. I've made the changes you've suggested. Apologies that it took me some time to act on these suggestions. Hopefully the changes should be good to go!

**Review by fortypercenttitanium — APPROVED** [review]
Excellent work @anupamvamsi .

**Pull request merged**